contents/blog/how-we-made-feature-flags-even-faster-and-even-more-reliable.md

```diff
@@ -0,0 +1,120 @@
+
+The language's type system makes it much harder to write buggy code, which matters when you're building infrastructure that thousands of applications depend on. We also had confidence from our capture service rewrite, where Rust had already demonstrated massive efficiency gains.
+
+Most importantly, Rust gave us proper code-level timeout primitives, eliminating our problematic dependency on PgBouncer-level settings. No more relying on external connection pooling for reliability.
+
+### Moving evaluation logic out of the database
+
+The biggest architectural change was moving flag condition evaluation from SQL to application code. Previously, we were doing complex property matching inside database queries, which required expensive joins and data transfers.
+
+Now we:
+
+1. Fetch all relevant person properties in a single query at the start of each request
+2. Evaluate all flag conditions in memory, in parallel
+3. Return results without additional database round-trips
+
+This front-loads the database work while moving CPU-intensive evaluation logic to where it belongs—in application code that can leverage multiple cores effectively. It also means we're transferring less data and can cache person properties more effectively.
+
+### Cohort caching
+
+Cohort-based flags were some of our slowest queries because they required complex joins to determine membership. We now cache cohort membership at the application level rather than computing it on-demand for every flag evaluation.
+
+This eliminated one of our biggest performance bottlenecks and made cohort-based flags just as fast as simple property-based flags.
+
+### Simplified architecture
+
+We removed PgBouncer from the deployment entirely. Fewer moving parts means fewer things that can break, and much simpler debugging when they do.
+
+The new service connects directly to the database with proper connection pooling built into the Rust application itself, giving us much better control over timeouts and error handling.
+
+## Results
+
+The improvements were immediate and substantial:
+
+**Performance gains:**
+
+- p99 latency: 400ms → 71ms (82% improvement)
+- p95 latency: 138ms → 37ms (73% improvement)  
+- p90 latency: 81ms → 22ms (73% improvement)
+- p50 latency: 16ms → 4.4ms (72% improvement)
+
+These aren't just synthetic benchmarks—these are real-world improvements under production load with the same traffic patterns.
+
+**Cost reduction:**
+
+The efficiency gains were similarly ridiculous. Our Django service required ~300 pods to handle ~500k requests per minute at a cost of ~$8k/month. The new Rust service handles the same ~500k req/min using only 60 pods, costing around ~$1k/month.
+
+That's an 87% reduction in compute costs while serving identical traffic. The new service uses just 13% of the compute resources compared to the previous implementation.
```

Comment:
similarly here; I can include some graphs of pod utilization + cost savings

Reply:
same comment as [here](https://github.com/PostHog/posthog.com/pull/12298/files#r2393491139), because the services are separate, there's no cool line go down image, it's just tables with different numbers in them. I don't think it's any punchier than the text.